<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>IFAA</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">IFAA</h1>



<p>IFAA offers a robust approach to make inference on the association of
covariates with the absolute abundance (AA) of microbiome in an
ecosystem. It can be also directly applied to relative abundance (RA)
data to make inference on AA because the ratio of two RA is equal ratio
of their AA. This algorithm can estimate and test the associations of
interest while adjusting for potential confounders. High-dimensional
covariates are handled with regularization. The estimates of this method
have easy interpretation like a typical regression analysis.
High-dimensional covariates are handled with regularization and it is
implemented by parallel computing. False discovery rate is automatically
controlled by this approach. Zeros do not need to be imputed by a
positive value for the analysis. The IFAA package also offers the
‘MZILN’ function for estimating and testing associations of abundance
ratios with covariates.</p>
<p>To model the association, the following equation is used: <span class="math display">\[
\log(\mathcal{Y}_i^k)|\mathcal{Y}_i^k&gt;0=\beta^{0k}+X_i^T\beta^k+W_i^T\gamma^k+Z_i^Tb_i+\epsilon_i^k,\hspace{0.2cm}k=1,...,K+1,
\]</span> where</p>
<ul>
<li><p><span class="math inline">\(\mathcal{Y}_i^k\)</span> is the AA of
taxa <span class="math inline">\(k\)</span> in subject <span class="math inline">\(i\)</span> in the entire ecosystem.</p></li>
<li><p><span class="math inline">\(X_i\)</span> is the covariate
matrix.</p></li>
<li><p><span class="math inline">\(W_i\)</span> is the confounder
matrix.</p></li>
<li><p><span class="math inline">\(Z_i\)</span> is the design matrix for
random effects.</p></li>
<li><p><span class="math inline">\(\beta^k\)</span> is the regression
coefficients that will be estimated and tested with the
<code>IFAA()</code> function.</p></li>
</ul>
<p>The challenge in microbiome analysis is that we can not oberve <span class="math inline">\(\mathcal{Y}_i^k\)</span>. What is observed is its
small proportion: <span class="math inline">\(Y_i^k=C_i\mathcal{Y}^k_i\)</span> where <span class="math inline">\(C_i\)</span> is an unknown number between 0 and 1
that denote the observed proportion. The IFAA method successfuly
addressed this challenge.</p>
<div id="package-installation" class="section level2">
<h2>Package installation</h2>
<p>To install, type the following command in R console:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;IFAA&quot;</span>, <span class="at">repos =</span> <span class="st">&quot;http://cran.us.r-project.org&quot;</span>)</span></code></pre></div>
<p>The package could be also installed from GitHub using the following
code:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(devtools)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;gitlzg/IFAA&quot;</span>)</span></code></pre></div>
</div>
<div id="input-for-ifaa-function" class="section level2">
<h2>Input for IFAA() function</h2>
<p>Most of the time, users just need to feed the first three inputs to
the function: <code>experiment_dat</code>, <code>testCov</code> and
<code>ctrlCov</code>. All other inputs can just take their default
values. Below are all the inputs of the functions</p>
<ul>
<li><p><code>experiment_dat</code>: A SummarizedExperiment object
containing microbiome data and covarites (see example on how to create a
SummarizedExperiment object). The microbiome data can be absolute
abundance or relative abundance with each column per sample and each row
per taxon/OTU/ASV (or any other unit). No imputation is needed for
zero-valued data points. The covarites data contains covariates and
confounders with each row per sample and each column per variable. The
covarites data has to be numeric or binary.</p></li>
<li><p><code>testCov</code>: Covariates that are of primary interest for
testing and estimating the associations. It corresponds to <span class="math inline">\(X_i\)</span> in the equation. Default is
<code>NULL</code> which means all covariates are
<code>testCov</code>.</p></li>
<li><p><code>ctrlCov</code>: Potential confounders that will be adjusted
in the model. It corresponds to <span class="math inline">\(W_i\)</span>
in the equation. Default is <code>NULL</code> which means all covariates
except those in <code>testCov</code> are adjusted as
confounders.</p></li>
<li><p><code>sampleID</code>: Name of the sample ID variable in the
data. This is not required. Default is NULL.</p></li>
<li><p><code>testMany</code>: This takes logical value <code>TRUE</code>
or <code>FALSE</code>. If <code>TRUE</code>, the <code>testCov</code>
will contain all the variables in <code>CovData</code> provided
<code>testCov</code> is set to be <code>NULL</code>. The default value
is <code>TRUE</code> which does not do anything if <code>testCov</code>
is not <code>NULL</code>.</p></li>
<li><p><code>ctrlMany</code>: This takes logical value <code>TRUE</code>
or <code>FALSE</code>. If <code>TRUE</code>, all variables except
<code>testCov</code> are considered as control covariates provided
<code>ctrlCov</code> is set to be <code>NULL</code>. The default value
is <code>FALSE</code>.</p></li>
<li><p><code>nRef</code>: The number of randomly picked reference taxa
used in phase 1. Default number is <code>40</code>.</p></li>
<li><p><code>nRefMaxForEsti</code>: The maximum number of final
reference taxa used in phase 2. The default is <code>2</code>.</p></li>
<li><p><code>refTaxa</code>: A vector of taxa names. These are reference
taxa specified by the user to be used in phase 1 if the user believe
these taxa are indepenent of the covariates. If the number of reference
taxa is less than ‘nRef’, the algorithm will randomly pick extra
reference taxa to make up ‘nRef’. The default is <code>NULL</code> since
the algorithm will pick reference taxa randomly.</p></li>
<li><p><code>adjust_method</code>: The adjusting method used for p value
adjustment. Default is “BY” for dependent FDR adjustment. It can take
any adjustment method for p.adjust function in R.</p></li>
<li><p><code>fdrRate</code>: The false discovery rate for identifying
taxa/OTU/ASV associated with <code>testCov</code>. Default is
<code>0.15</code>.</p></li>
<li><p><code>paraJobs</code>: If <code>sequentialRun</code> is
<code>FALSE</code>, this specifies the number of parallel jobs that will
be registered to run the algorithm. If specified as <code>NULL</code>,
it will automatically detect the cores to decide the number of parallel
jobs. Default is <code>NULL</code>.</p></li>
<li><p><code>bootB</code>: Number of bootstrap samples for obtaining
confidence interval of estimates in phase 2 for the high dimensional
regression. The default is <code>500</code>.</p></li>
<li><p><code>standardize</code>: This takes a logical value
<code>TRUE</code> or <code>FALSE</code>. If <code>TRUE</code>, the
design matrix for X will be standardized in the analyses and the
results. Default is <code>FALSE</code>.</p></li>
<li><p><code>sequentialRun</code>: This takes a logical value
<code>TRUE</code> or <code>FALSE</code>. Default is <code>FALSE</code>.
This argument could be useful for debug.</p></li>
<li><p><code>refReadsThresh</code>: The threshold of proportion of
non-zero sequencing reads for choosing the reference taxon in phase 2.
The default is <code>0.2</code> which means at least 20% non-zero
sequencing reads.</p></li>
<li><p><code>taxDropThresh</code>: The threshold of number of non-zero
sequencing reads for each taxon to be dropped from the analysis. The
default is <code>0</code> which means taxon without any sequencing reads
will be dropped from the analysis.</p></li>
<li><p><code>SDThresh</code>: The threshold of standard deviations of
sequencing reads for been chosen as the reference taxon in phase 2. The
default is <code>0.05</code> which means the standard deviation of
sequencing reads should be at least <code>0.05</code> in order to be
chosen as reference taxon.</p></li>
<li><p><code>SDquantilThresh</code>: The threshold of the quantile of
standard deviation of sequencing reads, above which could be selected as
reference taxon. The default is <code>0</code>.</p></li>
<li><p><code>balanceCut</code>: The threshold of the proportion of
non-zero sequencing reads in each group of a binary variable for
choosing the final reference taxa in phase 2. The default number is
<code>0.2</code> which means at least 20% non-zero sequencing reads in
each group are needed to be eligible for being chosen as a final
reference taxon.</p></li>
<li><p><code>seed</code>: Random seed for reproducibility. Default is
<code>1</code>. It can be set to be NULL to remove seeding.</p></li>
</ul>
</div>
<div id="output-for-ifaa-function" class="section level2">
<h2>Output for IFAA() function</h2>
<p>A list containing 2 elements</p>
<ul>
<li><p><code>full_results</code>: The main results for IFAA containing
the estimation and testing results for all associations between all taxa
and all test covariates in <code>testCov</code>. It is a dataframe with
each row representing an association, and eight columns named as
“taxon”, “cov”, “estimate”, “SE.est”, “CI.low”, “CI.up”, “adj.p.value”,
and “sig_ind”. The columns correspond to taxon name, covariate name,
association estimates, standard error estimates, lower bound and upper
bound of the 95% confidence interval, adjusted p value, and the
indicator showing whether the association is significant after multiple
testing adjustment.</p></li>
<li><p><code>metadata</code>: The metadata is a list containing the
following items: <code>covariatesData</code>: A dataset containing
covariates and confounders used in the analyses.
<code>final_ref_taxon</code>: The final 2 reference taxa used for
analysis. <code>ref_taxon_count</code>: The counts of selection for the
associations of all taxa with test covariates in Phase 1.
<code>ref_taxon_est</code>: The average magnitude estimates for the
associations of all taxa with test covariates in Phase 1.
<code>totalTimeMins</code>: Total time used for the entire analysis.
<code>seed</code>: The seed used for the analysis for reproducibility.
<code>fdrRate</code>: FDR rate used for the analysis.
<code>adjust_method</code>: Multiple testing adjust method used for the
analysis.</p></li>
</ul>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>The example datasets <code>dataM</code> and <code>dataC</code> are
included in the package. The input should be a SummarizedExperiment
object, which could be constructed as:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(IFAA)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(SummarizedExperiment))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">## load the example microbiome data. This could be relative abundance or </span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="do">## absolute abundance data. If you have a csv or tsv file for the microbiome data, </span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="do">## you can use read.csv() function or read.table() function in R to read the </span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="do">## data file into R.</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataM)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dataM)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 40 61</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>dataM[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   id rawCount1 rawCount2 rawCount3 rawCount4 rawCount5 rawCount6 rawCount7</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  1         4        49         2         0       360       222         4</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  2         0        20        14         0        86       211         5</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  3         3         0         3         7         0        57         0</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4  4         9        18         5        31        42        58         8</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5  5         0         2         1        19        15        67         6</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="do">## load the example covariates data. If you have a csv or tsv file for the </span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="do">## covariates data, you can use read.csv() function or read.table() function </span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="do">## in R to read the data file into R.</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataC)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dataC)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 40  4</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>dataC[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ]</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   id        v1        v2        v3</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  1  58.06969 -49.90376 -15.30643</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  2  25.96522 -68.58894 -23.10992</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  3 193.71625 124.40186 119.56747</span></span></code></pre></div>
<p>Both the microbiome data <code>dataM</code> and the covariates data
<code>dataC</code> contain 40 samples (i.e., 40 rows).</p>
<ul>
<li><p><code>dataM</code> contains 60 taxa with absolute abundances and
these are gut microbiome.</p></li>
<li><p><code>dataC</code> contains 3 covariates.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Merge the microbiome data and covariate data by id to avoid unmatching observations. </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>data_merged<span class="ot">&lt;-</span><span class="fu">merge</span>(dataM,dataC,<span class="at">by=</span><span class="st">&quot;id&quot;</span>,<span class="at">all=</span><span class="cn">FALSE</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Seperate microbiome data and covariate data, drop id variable from microbiome data</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>dataM_sub<span class="ot">&lt;-</span>data_merged[,<span class="fu">colnames</span>(dataM)[<span class="sc">!</span><span class="fu">colnames</span>(dataM)<span class="sc">%in%</span><span class="fu">c</span>(<span class="st">&quot;id&quot;</span>)]]</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>dataC_sub<span class="ot">&lt;-</span>data_merged[,<span class="fu">colnames</span>(dataC)]</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Create a SummarizedExperiment object for inputs </span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>test_dat<span class="ot">&lt;-</span><span class="fu">SummarizedExperiment</span>(<span class="at">assays=</span><span class="fu">list</span>(<span class="at">counts=</span><span class="fu">t</span>(dataM_sub)), <span class="at">colData=</span>dataC_sub)</span></code></pre></div>
<p>Next we analyze the data to test the association between microbiome
and the variable <code>&quot;v1&quot;</code> while adjusting for the variables
(potential confounders) <code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">IFAA</span>(<span class="at">experiment_dat =</span> test_dat,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">testCov =</span> <span class="fu">c</span>(<span class="st">&quot;v1&quot;</span>),</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">ctrlCov =</span> <span class="fu">c</span>(<span class="st">&quot;v2&quot;</span>,<span class="st">&quot;v3&quot;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">sampleID =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">fdrRate =</span> <span class="fl">0.05</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data dimensions (after removing missing data if any):</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 40 samples</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 60 taxa/OTU/ASV</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 testCov variables in the analysis</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; These are the testCov variables:</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; v1</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 ctrlCov variables in the analysis</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; These are the ctrlCov variables:</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; v2, v3</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0 binary covariates in the analysis</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 25.71 percent of microbiome sequencing reads are zero</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Start Phase 1 analysis</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 parallel jobs are registered for analyzing 40 reference taxa in Phase 1</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 33 percent of phase 1 analysis has been done and it took 0.62 minutes</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 parallel jobs are registered for analyzing 20 reference taxa in Phase 1</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 67 percent of phase 1 analysis has been done and it took 1.05 minutes</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 parallel jobs are registered for analyzing 20 reference taxa in Phase 1</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 100 percent of phase 1 analysis has been done and it took 1.47 minutes</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Start Phase 2 parameter estimation</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 50 percent of Phase 2 is done and it took 0.019 minutes</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Entire Phase 2 parameter estimation done and took 0.04 minutes.</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The entire analysis took 1.51 minutes</span></span></code></pre></div>
<p>In this example, we are only interested in testing the associations
with <code>&quot;v1&quot;</code> which is why <code>testCov=c(&quot;v1&quot;)</code>. The
variables <code>&quot;v2&quot; and &quot;v3&quot;</code> are adjusted as potential
confounders in the analyses. The final analysis results are saved in the
list <code>full_result</code> and the significant results can be
extracted as follows:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>summary_res<span class="ot">&lt;-</span>results<span class="sc">$</span>full_result</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sig_results<span class="ot">&lt;-</span><span class="fu">subset</span>(summary_res,sig_ind<span class="sc">==</span>T)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>sig_results</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; DataFrame with 3 rows and 8 columns</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         taxon         cov  estimate     SE.est    CI.low     CI.up adj.p.value</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   &lt;character&gt; &lt;character&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  rawCount18          v1 0.0248316 0.00523955 0.0145620 0.0351011 1.92686e-04</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  rawCount36          v1 0.0285573 0.00564242 0.0174982 0.0396165 5.61399e-05</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  rawCount41          v1 0.0300483 0.00515387 0.0199467 0.0401499 1.49157e-06</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     sig_ind</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   &lt;numeric&gt;</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1         1</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2         1</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3         1</span></span></code></pre></div>
<p>The results found three taxa <code>&quot;rawCount18&quot;</code>,
<code>&quot;rawCount36&quot;</code>, <code>&quot;rawCount41&quot;</code> associated with
<code>&quot;v1&quot;</code> while adjusting for <code>&quot;v2&quot; and &quot;v3&quot;</code>. The
regression coefficients and their 95% confidence intervals are provided.
These coefficients correspond to <span class="math inline">\(\beta^k\)</span> in the model equation.</p>
<p>The interpretation is that</p>
<ul>
<li>Every unit increase in <code>&quot;v1&quot;</code> is associated with
approximately 2.5% increase in the absolute abundance of
<code>&quot;rawCount18&quot;</code>, approximately 2.9% increase in the absolute
abundance of <code>&quot;rawCount36&quot;</code>, and approximately 3.0% increase
in the absolute abundance of <code>&quot;rawCount41&quot;</code> in the entire gut
ecosystem.</li>
</ul>
</div>
<div id="mziln-function" class="section level2">
<h2>MZILN() function</h2>
<p>The IFAA package can also implement the Multivariate Zero-Inflated
Logistic Normal (MZILN) regression model for estimating and testing the
association of abundance ratios with covariates. The
<code>MZILN()</code> function estimates and tests the associations of
user-specified abundance ratios with covariates. When the denominator
taxon of the ratio is independent of the covariates, ‘MZILN()’ should
generate similar results as ‘IFAA()’. The regression model of ‘MZILN()’
can be expressed as follows: <span class="math display">\[
\log\bigg(\frac{\mathcal{Y}_i^k}{\mathcal{Y}_i^{K+1}}\bigg)|\mathcal{Y}_i^k&gt;0,\mathcal{Y}_i^{K+1}&gt;0=\alpha^{0k}+\mathcal{X}_i^T\alpha^k+\epsilon_i^k,\hspace{0.2cm}k=1,...,K,
\]</span> where</p>
<ul>
<li><p><span class="math inline">\(\mathcal{Y}_i^k\)</span> is the AA of
taxa <span class="math inline">\(k\)</span> in subject <span class="math inline">\(i\)</span> in the entire ecosystem.</p></li>
<li><p><span class="math inline">\(\mathcal{Y}_i^{K+1}\)</span> is the
reference taxon (specified by user).</p></li>
<li><p><span class="math inline">\(\mathcal{X}_i\)</span> is the
covariate matrix for all covariates including confounders.</p></li>
<li><p><span class="math inline">\(\alpha^k\)</span> is the regression
coefficients that will be estimated and tested.</p></li>
</ul>
</div>
<div id="input-for-mziln-function" class="section level2">
<h2>Input for MZILN() function</h2>
<p>Most of the time, users just feed the first three inputs to the
function: <code>experiment_dat</code>, <code>refTaxa</code> and
<code>allCov</code>. All other inputs can just take their default
values. All the inputs for ‘MZILN()’ are:</p>
<ul>
<li><p><code>experiment_dat</code>: A SummarizedExperiment object
containing microbiome data and covarites (see example on how to create a
SummarizedExperiment object). The microbiome data can be absolute
abundance or relative abundance with each column per sample and each row
per taxon/OTU/ASV (or any other unit). No imputation is needed for
zero-valued data points. The covarites data contains covariates and
confounders with each row per sample and each column per variable. The
covarites data has to be numeric or binary.</p></li>
<li><p><code>refTaxa</code>: Denominator taxa names specified by the
user for the targeted ratios. This could be a vector of names.</p></li>
<li><p><code>allCov</code>: All covariates of interest (including
confounders) for estimating and testing their associations with the
targeted ratios. Default is ‘NULL’ meaning that all covariates in
covData are of interest.</p></li>
<li><p><code>sampleID</code>: Name of the sample ID variable in the
data. This is not required. Default is NULL.</p></li>
<li><p><code>adjust_method</code>: The adjusting method for p value
adjustment. Default is “BY” for dependent FDR adjustment. It can take
any adjustment method for p.adjust function in R.</p></li>
<li><p><code>fdrRate</code> The false discovery rate for identifying
ratios associated with <code>allCov</code>. Default is
<code>0.15</code>.</p></li>
<li><p><code>paraJobs</code>: If <code>sequentialRun</code> is
<code>FALSE</code>, this specifies the number of parallel jobs that will
be registered to run the algorithm. If specified as <code>NULL</code>,
it will automatically detect the cores to decide the number of parallel
jobs. Default is <code>NULL</code>.</p></li>
<li><p><code>bootB</code>: Number of bootstrap samples for obtaining
confidence interval of estimates for the high dimensional regression.
The default is <code>500</code>.</p></li>
<li><p><code>taxDropThresh</code>: The threshold of number of non-zero
sequencing reads for each taxon to be dropped from the analysis. The
default is <code>0</code> which means taxon without any sequencing reads
will be dropped from the analysis.</p></li>
<li><p><code>standardize</code>: This takes a logical value
<code>TRUE</code> or <code>FALSE</code>. If <code>TRUE</code>, the
design matrix for X will be standardized in the analyses and the
results. Default is <code>FALSE</code>.</p></li>
<li><p><code>sequentialRun</code>: This takes a logical value
<code>TRUE</code> or <code>FALSE</code>. Default is <code>TRUE</code>.
It can be set to be “FALSE” to increase speed if there are multiple taxa
in the argument ‘refTaxa’.</p></li>
<li><p><code>seed</code>: Random seed for reproducibility. Default is
<code>1</code>. It can be set to be NULL to remove seeding.</p></li>
</ul>
</div>
<div id="output-for-mziln-function" class="section level2">
<h2>Output for MZILN() function</h2>
<p>A list with two elements:</p>
<ul>
<li><p><code>full_results</code>: The main results for MZILN containing
the estimation and testing results for all associations between all taxa
ratios with refTaxan being the denominator and all covariates in
<code>allCov</code>. It is a dataframe with each row representing an
association, and ten columns named as “ref_tax”, “taxon”, “cov”,
“estimate”, “SE.est”, “CI.low”, “CI.up”, “adj.p.value”, “unadj.p.value”
and “sig_ind”. The columns correspond to the denominator taxon,
numerator taxon, covariate name, association estimates, standard error
estimates, lower bound and upper bound of the 95% confidence interval,
adjusted p value, and the indicator showing whether the association is
significant after multiple testing adjustment.</p></li>
<li><p><code>metadata</code>: The metadata is a list containing total
time used in minutes, random seed used, FDR rate, and multiple testing
adjustment method used.</p></li>
</ul>
</div>
<div id="examples" class="section level2">
<h2>Examples</h2>
<p>We use the same example data The example dataset as that for
illustrating the IFAA function. <code>dataM</code> and
<code>dataC</code> are included in the package. They could be accessed
by:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## load the example microbiome data. This could be relative abundance or </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="do">## absolute abundance data. If you have a csv or tsv file for the microbiome data, </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="do">## you can use read.csv() function or read.table() function in R to read the </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="do">## data file into R.</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataM)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dataM)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 40 61</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>dataM[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   id rawCount1 rawCount2 rawCount3 rawCount4 rawCount5 rawCount6 rawCount7</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  1         4        49         2         0       360       222         4</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  2         0        20        14         0        86       211         5</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  3         3         0         3         7         0        57         0</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4  4         9        18         5        31        42        58         8</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5  5         0         2         1        19        15        67         6</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="do">## load the example covariates data. If you have a csv or tsv file for the </span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="do">## covariates data, you can use read.csv() function or read.table() function </span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="do">## in R to read the data file into R.</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataC)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dataC)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 40  4</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>dataC[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ]</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   id        v1        v2        v3</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  1  58.06969 -49.90376 -15.30643</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  2  25.96522 -68.58894 -23.10992</span></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  3 193.71625 124.40186 119.56747</span></span></code></pre></div>
<p>Both the microbiome data <code>dataM</code> and the covariates data
<code>dataC</code> contain 40 samples (i.e., 40 rows).</p>
<ul>
<li><p><code>dataM</code> contains 60 taxa with absolute abundances and
these are gut microbiome.</p></li>
<li><p><code>dataC</code> contains 3 covariates.</p></li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## load the example microbiome data. This could be relative abundance or </span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="do">## absolute abundance data. If you have a csv or tsv file for the microbiome data, </span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="do">## you can use read.csv() function or read.table() function in R to read the </span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="do">## data file into R.</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>data_merged<span class="ot">&lt;-</span><span class="fu">merge</span>(dataM,dataC,<span class="at">by=</span><span class="st">&quot;id&quot;</span>,<span class="at">all=</span><span class="cn">FALSE</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="do">## load the covariates data. If you have a csv or tsv file for the covariates data, </span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="do">## you can use read.csv() function or read.table() function in R to read </span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="do">## the data file into R.</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>dataM_sub<span class="ot">&lt;-</span>data_merged[,<span class="fu">colnames</span>(dataM)[<span class="sc">!</span><span class="fu">colnames</span>(dataM)<span class="sc">%in%</span><span class="fu">c</span>(<span class="st">&quot;id&quot;</span>)]]</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>dataC_sub<span class="ot">&lt;-</span>data_merged[,<span class="fu">colnames</span>(dataC)]</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="do">## Create a SummarizedExperiment object for inputs </span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>test_dat<span class="ot">&lt;-</span><span class="fu">SummarizedExperiment</span>(<span class="at">assays=</span><span class="fu">list</span>(<span class="at">counts=</span><span class="fu">t</span>(dataM_sub)), <span class="at">colData=</span>dataC_sub)</span></code></pre></div>
<p>Next we analyze the data to test the associations between the ratio
“rawCount18/rawCount11” and all the three variables <code>&quot;v1&quot;</code>,
<code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code> in a multivariate model where
all <code>&quot;v1&quot;</code>, <code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code> are
independent variables simultaneously.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">MZILN</span>(<span class="at">experiment_dat=</span>test_dat,</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">refTaxa=</span><span class="fu">c</span>(<span class="st">&quot;rawCount11&quot;</span>),</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">allCov=</span><span class="fu">c</span>(<span class="st">&quot;v1&quot;</span>,<span class="st">&quot;v2&quot;</span>,<span class="st">&quot;v3&quot;</span>),</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">sampleID =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>),</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fdrRate=</span><span class="fl">0.05</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Data dimensions (after removing missing data if any):</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 40 samples</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 60 taxa/OTU/ASV</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 testCov variables in the analysis</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; These are the testCov variables:</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; v1, v2, v3</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0 ctrlCov variables in the analysis</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0 binary covariates in the analysis</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 25.71 percent of microbiome sequencing reads are zero</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Estimation done for the 1th denominator taxon: rawCount11 and it took 0.02 minutes</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The entire analysis took 0.02 minutes</span></span></code></pre></div>
<p>The full final analysis results can be extracted as follows:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>summary_res<span class="ot">&lt;-</span>results<span class="sc">$</span>full_results</span></code></pre></div>
<p>The results for the log-ratio of “rawCount18” over “rawCount11” can
extracted as follows:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>summary_res[summary_res<span class="sc">$</span>taxon<span class="sc">==</span><span class="st">&quot;rawCount18&quot;</span>,,drop<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; DataFrame with 3 rows and 10 columns</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       ref_tax       taxon         cov    estimate     SE.est      CI.low</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt;   &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  rawCount11  rawCount18          v1  0.02310369 0.00557079  0.01218495</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  rawCount11  rawCount18          v2  0.00260412 0.00317369 -0.00361632</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  rawCount11  rawCount18          v3 -0.00625053 0.00281432 -0.01176659</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;          CI.up adj.p.value unadj.p.value   sig_ind</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      &lt;numeric&gt;   &lt;numeric&gt;     &lt;numeric&gt; &lt;logical&gt;</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  0.034022440  0.00308539   0.000033643      TRUE</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  0.008824558  1.00000000   0.411913110     FALSE</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 -0.000734468  0.80559644   0.026352611     FALSE</span></span></code></pre></div>
<p>The regression coefficients and their 95% confidence intervals are
provided. These coefficients correspond to <span class="math inline">\(\alpha^k\)</span> in the model equation, and can
be interpreted as the associations between the covariates and log-ratio
of <code>&quot;rawCount18&quot;</code> over ‘“rawCount11”’.</p>
<p>The interpretation for the results is that</p>
<ul>
<li>Every unit increase in <code>&quot;v1&quot;</code> is associated with
approximately 2.3% increase in the abundance ratio of
<code>&quot;rawCount18&quot;</code> over <code>&quot;rawCount11&quot;</code> (while
controlling for <code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code>); Every unit
increase in <code>&quot;v2&quot;</code> is associated with approximately 0.26%
increase in the abundance ratio of <code>&quot;rawCount18&quot;</code> over
<code>&quot;rawCount11&quot;</code> (while controlling for <code>&quot;v1&quot;</code> and
<code>&quot;v3&quot;</code>), but not statistically significant; Every unit
increase in <code>&quot;v3&quot;</code> is associated with approximately -0.63%
decrease in the abundance ratio of <code>&quot;rawCount18&quot;</code> over
<code>&quot;rawCount11&quot;</code> (while controlling for <code>&quot;v1&quot;</code> and
<code>&quot;v2&quot;</code>), but not statistically significant.</li>
</ul>
<p>We can also extract all the ratios (with <code>&quot;rawCount11&quot;</code>
being the denominator taxon) that are significantly associated with any
of the covariates as follows:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(summary_res,sig_ind<span class="sc">==</span>T)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; DataFrame with 3 rows and 10 columns</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       ref_tax       taxon         cov  estimate     SE.est    CI.low     CI.up</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1  rawCount11  rawCount18          v1 0.0231037 0.00557079 0.0121849 0.0340224</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  rawCount11  rawCount36          v1 0.0296571 0.00589475 0.0181034 0.0412108</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3  rawCount11  rawCount41          v1 0.0256273 0.00546293 0.0149199 0.0363346</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   adj.p.value unadj.p.value   sig_ind</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     &lt;numeric&gt;     &lt;numeric&gt; &lt;logical&gt;</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 0.003085391   3.36430e-05      TRUE</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 0.000134166   4.87647e-07      TRUE</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 0.000373778   2.71711e-06      TRUE</span></span></code></pre></div>
<p>The interpretation for the results is that</p>
<ul>
<li>Every unit increase in <code>&quot;v1&quot;</code> is associated with
approximately 2.3% increase in the abundance ratio of
<code>&quot;rawCount18&quot;</code> over <code>&quot;rawCount11&quot;</code> (while
controlling for <code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code>), and it is
statistically significant; Every unit increase in <code>&quot;v1&quot;</code> is
also associated with approximately 3.0% increase in the abundance ratio
of <code>&quot;rawCount36&quot;</code> over <code>&quot;rawCount11&quot;</code> (while
controlling for <code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code>), and it is
statistically significant; Every unit increase in <code>&quot;v1&quot;</code> is
also associated with approximately 2.6% decrease in the abundance ratio
of <code>&quot;rawCount41&quot;</code> over <code>&quot;rawCount11&quot;</code> (while
controlling for <code>&quot;v2&quot;</code> and <code>&quot;v3&quot;</code>), and it is
statistically significant.</li>
</ul>
<p>Session Info</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; R version 4.2.1 (2022-06-23 ucrt)</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 10 x64 (build 19044)</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=English_United States.utf8 </span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=English_United States.utf8   </span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=English_United States.utf8</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                          </span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=English_United States.utf8    </span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [8] base     </span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] SummarizedExperiment_1.26.1 Biobase_2.56.0             </span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [3] GenomicRanges_1.48.0        GenomeInfoDb_1.32.2        </span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [5] IRanges_2.30.0              S4Vectors_0.34.0           </span></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [7] BiocGenerics_0.42.0         MatrixGenerics_1.8.0       </span></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [9] matrixStats_0.62.0          IFAA_1.0.7                 </span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] Rcpp_1.0.7             mvtnorm_1.1-3          lattice_0.20-45       </span></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [4] class_7.3-20           glmnet_4.1-3           digest_0.6.29         </span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [7] RhpcBLASctl_0.21-247.1 foreach_1.5.1          parallelly_1.30.0     </span></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [10] slam_0.1-49            R6_2.5.1               cellranger_1.1.0      </span></span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [13] sparsesvd_0.2          qlcMatrix_0.9.7        evaluate_0.14         </span></span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [16] rootSolve_1.8.2.3      e1071_1.7-11           httr_1.4.2            </span></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [19] zlibbioc_1.42.0        rlang_1.0.3            Exact_3.1             </span></span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [22] readxl_1.4.0           rstudioapi_0.13        data.table_1.14.2     </span></span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [25] jquerylib_0.1.4        Matrix_1.4-1           rmarkdown_2.12        </span></span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [28] mathjaxr_1.4-0         splines_4.2.1          stringr_1.4.0         </span></span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [31] RCurl_1.98-1.7         DelayedArray_0.22.0    proxy_0.4-27          </span></span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [34] compiler_4.2.1         xfun_0.29              speedglm_0.3-4        </span></span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [37] shape_1.4.6            DescTools_0.99.45      htmltools_0.5.2       </span></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [40] lmom_2.9               GenomeInfoDbData_1.2.8 MatrixExtra_0.1.10    </span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [43] expm_0.999-6           codetools_0.2-18       MASS_7.3-57           </span></span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [46] bitops_1.0-7           grid_4.2.1             jsonlite_1.7.2        </span></span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [49] magrittr_2.0.1         docopt_0.7.1           gld_2.6.4             </span></span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [52] cli_3.3.0              stringi_1.7.6          XVector_0.36.0        </span></span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [55] doParallel_1.0.16      bslib_0.3.1            boot_1.3-28           </span></span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [58] HDCI_1.0-2             iterators_1.0.13       tools_4.2.1           </span></span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [61] float_0.3-0            parallel_4.2.1         fastmap_1.1.0         </span></span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [64] survival_3.3-1         yaml_2.2.1             knitr_1.37            </span></span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [67] sass_0.4.0</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
